sparse/impl/KokkosSparse_gmres_numeric_impl.hpp

```diff
@@ -0,0 +1,391 @@
+        // Givens rotation and compute shortcut residual:
+        for (int i = 0; i < j; i++) {
+          ST tempVal    = CosVal_h(i) * H_h(i, j) + SinVal_h(i) * H_h(i + 1, j);
+          H_h(i + 1, j) = -karith::conj(SinVal_h(i)) * H_h(i, j) +
+                          CosVal_h(i) * H_h(i + 1, j);
+          H_h(i, j) = tempVal;
+        }
+        ST f  = H_h(j, j);
+        ST g  = H_h(j + 1, j);
+        MT f2 = karith::real(f) * karith::real(f) +
+                karith::imag(f) * karith::imag(f);
+        MT g2 = karith::real(g) * karith::real(g) +
+                karith::imag(g) * karith::imag(g);
+        ST fg2        = f2 + g2;
+        ST D1         = one / karith::sqrt(f2 * fg2);
+        CosVal_h(j)   = f2 * D1;
+        fg2           = fg2 * D1;
+        H_h(j, j)     = f * fg2;
+        SinVal_h(j)   = f * D1 * karith::conj(g);
+        H_h(j + 1, j) = zero;
+
+        GVec_h(j + 1) = GVec_h(j) * (-karith::conj(SinVal_h(j)));
+        GVec_h(j)     = GVec_h(j) * CosVal_h(j);
+        shortRelRes   = karith::abs(GVec_h(j + 1)) / nrmB;
+
+        if (verbose) {
+          std::cout << "Shortcut relative residual for iteration "
+                    << j + (cycle * m) << " is: " << shortRelRes << std::endl;
+        }
+        if (tmpNrm <= 1e-14 && shortRelRes >= tol) {
+          throw std::runtime_error(
+              "GMRES has experienced lucky breakdown, but the residual has not converged.\n\
+                                  Solver terminated without convergence.");
+        }
+        if (karith::isNan(ST(shortRelRes))) {
+          throw std::runtime_error(
+              "gmres: Relative residual is nan. Terminating solver.");
+        }
+
+        // If short residual converged, or time to restart, check true residual
+        if (shortRelRes < tol || j == m - 1) {
+          // Compute least squares soln with Givens rotation:
+          auto GLsSolnSub_h = Kokkos::subview(
+              GLsSoln_h, Kokkos::ALL,
+              0);  // Original view has rank 2, need a rank 1 here.
+          auto GVecSub_h = Kokkos::subview(GVec_h, Kokkos::make_pair(0, m));
+          Kokkos::deep_copy(GLsSolnSub_h,
+                            GVecSub_h);  // Copy LS rhs vec for triangle solve.
+          auto GLsSolnSub2_h = Kokkos::subview(
+              GLsSoln_h, Kokkos::make_pair(0, j + 1), Kokkos::ALL);
+          auto H_Sub_h = Kokkos::subview(H_h, Kokkos::make_pair(0, j + 1),
+                                         Kokkos::make_pair(0, j + 1));
+          {
+            // Hack to get around uninstantiated trsm for layoutleft
+            KokkosBlas::Impl::SerialTrsm_Invoke("L", "U", "N", "N", one,
+                                                H_Sub_h, GLsSolnSub2_h);
+          }
+          Kokkos::deep_copy(GLsSoln, GLsSoln_h);
+
+          // Update solution and compute residual with Givens:
+          VSub = Kokkos::subview(V, Kokkos::ALL, Kokkos::make_pair(0, j + 1));
```

Comment:
For future: Can possibly replace some of this with the BLAS Givens rotation impl.